programming_examples/matrix_multiplication/bf16/Makefile

```diff
@@ -16,27 +25,35 @@ run2x4: compile-kernel
 run2x2: compile-kernel
 	cd build && ${powershell} python3 ${srcdir}/run.py --herd-m 2 --herd-n 2 --m 512 --n 512 --k 512 --compile-mode compile-and-run
 
-run1x4: compile-kernel
-	cd build && ${powershell} python3 ${srcdir}/run.py --herd-m 1 --herd-n 4 --m 512 --n 512 --k 512 --compile-mode compile-and-run
-
-run4x1:
-	cd build && ${powershell} python3 ${srcdir}/run.py --herd-m 4 --herd-n 1 --m 512 --n 512 --k 512 --compile-mode compile-and-run
-
-AIEOPT_DIR = $(shell realpath $(dir $(shell which aie-opt))/..)
-WARNING_FLAGS = -Wno-parentheses -Wno-attributes -Wno-macro-redefined -Wno-empty-body
-PEANOWRAP2_FLAGS = -O2 -std=c++20 --target=aie2-none-unknown-elf ${WARNING_FLAGS} -DNDEBUG -I ${AIEOPT_DIR}/include
-
 compile-kernel:
 	mkdir -p build
-	@if [ -n "$$LLVM_AIE_INSTALL_DIR" ]; then \
-		echo "Using clang++ from LLVM_AIE_INSTALL_DIR"; \
-		$$LLVM_AIE_INSTALL_DIR/bin/clang++ ${PEANOWRAP2_FLAGS} -DBIT_WIDTH=8 -c ${srcdir}/mm.cc -o build/mm.o; \
-	elif command -v xchesscc_wrapper >/dev/null 2>&1; then \
-		echo "Using xchesscc_wrapper from PATH"; \
-		cd build && ${powershell} xchesscc_wrapper aie2 -c ${srcdir}/mm.cc -o mm.o; \
+	@if [ "$(BACKEND)" = "peano" ]; then \
+		if [ -x "$(PEANO_INSTALL_DIR)/bin/clang++" ]; then \
+			echo "Using clang++ from PEANO_INSTALL_DIR=$(PEANO_INSTALL_DIR)"; \
+			$(PEANO_INSTALL_DIR)/bin/clang++ ${PEANOWRAP2_FLAGS} -DBIT_WIDTH=8 -c ${srcdir}/mm.cc -o build/mm.o; \
+		else \
+			echo "Error: llvm-aie pip package not found or invalid."; \
+			exit 1; \
+		fi \
+	elif [ "$(BACKEND)" = "xchesscc" ]; then \
+		if command -v xchesscc_wrapper >/dev/null 2>&1; then \
+			echo "Using xchesscc_wrapper from PATH"; \
+			cd build && ${powershell} xchesscc_wrapper aie2 -c ${srcdir}/mm.cc -o mm.o; \
+		else \
+			echo "Error: xchesscc_wrapper not found in PATH."; \
+			exit 1; \
+		fi \
 	else \
-		echo "Error: Neither LLVM_AIE_INSTALL_DIR nor xchesscc_wrapper found." >&2; \
-		exit 1; \
+		if [ -x "$(PEANO_INSTALL_DIR)/bin/clang++" ]; then \
+			echo "Auto-selected peano backend"; \
+			$(PEANO_INSTALL_DIR)/bin/clang++ ${PEANOWRAP2_FLAGS} -DBIT_WIDTH=8 -c ${srcdir}/mm.cc -o build/mm.o; \
+		elif command -v xchesscc_wrapper >/dev/null 2>&1; then \
+			echo "Auto-selected xchesscc backend"; \
+			cd build && ${powershell} xchesscc_wrapper aie2 -c ${srcdir}/mm.cc -o mm.o; \
+		else \
+			echo "Error: no usable backend found."; \
+			exit 1; \
+		fi \
```

Comment:
Is this `compile-kernel` rule the same everywhere? maybe a `makefile.common` is appropriate.